File: matlab/task1_2.m
% 1.2 Exibição de bandas individuais (R, G e B) como imagens
% monocromáticas ou coloridas (em tons de R, G ou B, respectivamente)
function task1_2()
    assetsDir = ['..' filesep 'assets' filesep];
    filePath = strcat(assetsDir, 'lenna.png');
    
    originalImage = imread(filePath);
    
    [redCh, greenCh, blueCh] = separateChannels(originalImage);
    
    zeroes = zeros(size(originalImage, 1), size(originalImage, 2));
    [redColor, greenColor, blueColor] = getColoredImage(redCh, greenCh, blueCh, zeroes);
    
    concatImage = cat(3, redCh, greenCh, blueCh);

    % Ploting
    fig = figure(1);
    set (fig, 'Units', 'normalized', 'Position', [0,0,1,1]);
    
    subplot(2,4,1), imshow(originalImage), title('Original image')

    subplot(2,4,2), imshow(redCh), title('Red channel mono')
    subplot(2,4,3), imshow(greenCh), title('Green channel mono')
    subplot(2,4,4), imshow(blueCh), title('Blue channel mono')

    subplot(2,4,5), imshow(redColor), title('Red channel')
    subplot(2,4,6), imshow(greenColor), title('Green channel')
    subplot(2,4,7), imshow(blueColor), title('Blue channel')

    subplot(2,4,8), imshow(concatImage), title('Back to original image')
    
end

function [redCh, greenCh, blueCh] = separateChannels(input)
    redCh = input(:,:,1); 
    greenCh = input(:,:,2);
    blueCh = input(:,:,3);
end

function [justRed, justGreen, justBlue] = getColoredImage(redCh, greenCh, blueCh, zeroes)    
    justRed = cat(3, redCh, zeroes, zeroes);
    justGreen = cat(3, zeroes, greenCh, zeroes);
    justBlue = cat(3, zeroes, zeroes, blueCh);
end

File: python/task1_2.py
# coding=UTF-8
# 1.2 Exibição de bandas individuais (R, G e B) como imagens
# monocromáticas ou coloridas (em tons de R, G ou B, respectivamente)

import numpy as np
import utils


def main():
    image = utils.load_image('lenna.png')
    utils.display_single_image('Original', image)

    blue_mono, green_mono, red_mono = utils.split_channels(image)
    utils.display_single_image('Blue channel mono', blue_mono)
    utils.display_single_image('Green channel mono', green_mono)
    utils.display_single_image('Red channel mono', red_mono)

    zeroes = np.zeros(image.shape[0:2], dtype='uint8')

    blue = utils.merge_channels(blue_mono, zeroes, zeroes)
    green = utils.merge_channels(zeroes, green_mono, zeroes)
    red = utils.merge_channels(zeroes, zeroes, red_mono)
    utils.display_single_image('Blue channel', blue)
    utils.display_single_image('Green channel', green)
    utils.display_single_image('Red channel', red)

    back_to_original = utils.merge_channels(blue_mono, green_mono, red_mono)
    utils.display_single_image('Back to Original Channel', back_to_original)

    utils.wait_key_and_destroy_windows()


if __name__ == "__main__":
    main()